p2p/nat.py

```diff
@@ -0,0 +1,178 @@
+import asyncio
+from concurrent.futures import (
+    ThreadPoolExecutor,
+)
+import ipaddress
+from typing import (
+    AsyncGenerator,
+    NamedTuple,
+)
+from urllib.parse import urlparse
+
+from p2p.cancel_token import (
+    CancelToken,
+)
+from p2p.exceptions import (
+    OperationCancelled,
+)
+import netifaces
+from p2p.service import BaseService
+import upnpclient
+
+
+# UPnP discovery can take a long time, so use a loooong timeout here.
+UPNP_DISCOVER_TIMEOUT_SECONDS = 30
+
+
+class PortMapping(NamedTuple):
+    internal: str  # of the form "192.2.3.4:56"
+    external: str  # of the form "192.2.3.4:56"
+
+
+class NoInternalAddressMatchesDevice(Exception):
```

Comment:
Is there a reason this isn't in `p2p.exceptions`?

Reply:
Everything in `p2p.exceptions` seems to be related to peers, and this doesn't really have anything to do with it. Maybe `nat` should be moved to `trinity`.

Reply:
I think it is safe to assume anyone who wants to setup an app which uses the DevP2P protocol will want this functionality so `p2p` seems more appropriate.